internal/mesh/internal/controllers/gatewayproxy/controller.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package gatewayproxy
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/controller/dependency"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/gatewayproxy/builder"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/gatewayproxy/fetcher"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecarproxy/cache"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// ControllerName is the name for this controller. It's used for logging or status keys.
+const ControllerName = "consul.io/gateway-proxy-controller"
+
+// Controller is responsible for triggering reconciler for watched resources
+func Controller(cache *cache.Cache) *controller.Controller {
```

Comment:
nit: should this be named `NewController`?

Reply:
I intentionally followed the patterns from `sidecarproxy.Controller` - which uses `Controller()` - so that it will be easier for devs to hop between or them if needed 

https://github.com/hashicorp/consul/blob/010bf533d11ac29a5632659cd2af2c5120a11f8e/internal/mesh/internal/controllers/sidecarproxy/controller.go#L31-L101